### DISCUSSION TITLE
some questions abount padding, combining panels, refresh

### DISCUSSION THREAD

**Discussion by nschloe** [Textualize/textual#117] [Q&A]
I started playing around with Textual yesterday and I though I make a little system monitor. I've gotten as far as 

![screenshot](https://user-images.githubusercontent.com/181628/134318174-1d055315-b6e0-4a22-8d94-b693aca42a13.png)

(compare [bpytop](https://github.com/aristocratos/bpytop)). There are a few things I don't know how to do (a mixture of Texual/Rich questions):

- Tables without borders? I'd like to eliminate the lines in the process view, but couldn't find which box style does that.
- Eliminate padding? The proc view has inner padding to the table, how to remove it?
- The cpu widget consists of a panel containing a panel. I would like to add stuff to the left of the inner panel, ideally without padding. I haven't found how to combine Panels/Aligns yet.
- Render frequency. I noticed that `self.render` is called more often than prescribed by `self.set_interval(2.0, self.refresh)`. Is that normal? I work around this by using `self.set_interval(2.0, self.collect_data)`, where `self.collect_data` calls `self.refresh()` at the end, but I'm not sure if this is intended.

**Comment by willmcgugan** [comment]
> Tables without borders? I'd like to eliminate the lines in the process view, but couldn't find which box style does that.

You can set `box=None` to remove lines entirely. 

> Eliminate padding? The proc view has inner padding to the table, how to remove it?

I think that padding is the edge of the table, which you can disable with `show_edge=False`

> The cpu widget consists of a panel containing a panel. I would like to add stuff to the left of the inner panel, ideally without padding. I haven't found how to combine Panels/Aligns yet.

You can put them in a table with no borders (use Table.grid()) or a Columns object if you want it to auto fit the available width.

> Render frequency. I noticed that self.render is called more often than prescribed by self.set_interval(2.0, self.refresh). Is that normal? I work around this by using self.set_interval(2.0, self.collect_data), where self.collect_data calls self.refresh() at the end, but I'm not sure if this is intended.

How are you noticing that? Widgets can refresh independently, so you may see more events than every 2 seconds.

Looks very cool BTW!

**Reply by nschloe** [reply]
> You can set box=None to remove lines entirely.

Perfect. ~~Is there a way to have one space between columns, too?~~

> I think that padding is the edge of the table, which you can disable with show_edge=False

:heavy_check_mark: 

> You can put them in a table with no borders (use Table.grid()) or a Columns object if you want it to auto fit the available width.

Is it possible to adapt the vertical alignment of cell content somehow? _Edit:_ [Moved to rich](https://github.com/willmcgugan/rich/discussions/1524)

> How are you noticing that? Widgets can refresh independently, so you may see more events than every 2 seconds.

It only happens if multiple widgets are present. Check this:

```python
from textual.app import App
from textual.widget import Widget

from rich.panel import Panel
from rich import box


class RenderCounter(Widget):
    def on_mount(self):
        self.k = [0]
        self.set_interval(2.0, self.refresh)

    def render(self) -> Panel:
        self.k.append(self.k[-1] + 1)
        return Panel(
            " ".join([str(item) for item in self.k]),
            box=box.SQUARE,
        )


class CounterApp(App):
    async def on_mount(self) -> None:
        await self.view.dock(RenderCounter(), size=10)
        await self.view.dock(RenderCounter(), size=10)
        await self.view.dock(RenderCounter(), size=10)
        await self.view.dock(RenderCounter(), size=10)
        await self.view.dock(RenderCounter(), size=10)


CounterApp.run()
```

I would expect all widgets to refresh once every two seconds. The first one does, but all others refresh twice. Also, they all start off with two refreshes to begin with.

![screenshot](https://user-images.githubusercontent.com/181628/134508113-7ec5c93a-cc25-4112-a058-f5a7999d2166.png)

**Reply by bclements** [reply]
@nschloe mind if I get a copy of your code for the view you show in our OP way up the top of the thread?? I'm trying to build something similar.

**Reply by mharig** [reply]
@bclements:
This is a rather old thread. You will find the code in nschloe/tiptop.

**Reply by nschloe** [reply]
Yup, it's tiptop, licensed under MIT.
